ytt-002 Accept plain YAML for Data Values

pivotaljohn opened the pull request:
A mode of accepting Data Values into a `ytt` invocation stripped of any requirements to understand `ytt` concepts or syntax.

jtigger committed:
Initial draft of 003-raw-data-values

pivotaljohn committed:
Fixed broken internal link in 003-raw-data-values

ewrenn8 commented on proposals/ytt/003-raw-data-values/README.md:
```diff
@@ -0,0 +1,283 @@
+          paths:
+          - "config.yml"
+          - "values.yml"
+          fileMarks:
```
We may want to provide some syntactic sugar for this and instead have kapp-controller handle creation of the flag, instead of the user providing a more raw form in the spec. 
e.g.
```yaml
...
template:
      - ytt:
          paths:
          - "config.yml"
          - "values.yml"
          valuesFrom:
          - secretRef:
              name: secret-name
          - configMapRef:
              name: cfgmap-name
          - path: values/shared.yml       
...
```
This would make the apis for providing values to templating steps more consistent since helm already uses the valuesFrom key for specifying values. We will write up an issue proposing proposing how we will integrate with this on kapp-controller's end and can link it here.

pivotaljohn replied:
Excellent. I'll remove anything that looks like it's attempting to specify what downstream tooling presents to end-users. 👍🏼

ewrenn8 commented on proposals/ytt/003-raw-data-values/README.md:
```diff
@@ -0,0 +1,283 @@
+
+`ytt` also sports the `--file...` family of command-line flags.
+
+Another option would be to extend that group of flags with `--file-data-values`.
```
+1 to avoiding this. We already have a flag `--data-value-file` and the combination of these flags could get pretty confusing.

pivotaljohn replied:
Thanks for the review!

jtigger committed:
Remove unnecessary specs on kapp-controller

- added rationale for rejecting `--file-data-values` flag.

pivotaljohn committed:
Mark proposal as "in review"

jtigger committed:
Follow emerging convention: one sequence per tool

cppforlife commented on proposals/ytt/002-raw-data-values/README.md:
```diff
@@ -0,0 +1,254 @@
+- "typing" a file, i.e. marking it as containing YAML documents that are Data Values;
+- where in the ordering a file is processed.
+
+Furthermore, there already exists a flag named `--data-value-file`. This proposed flag would be confusingly similar.
```
i imagine in this alternative scenario we would call it something like `--data-values-file` since it represents multiple data values coming from a file (similar to --data-values-env).

cppforlife replied:
i do like this approach as well. this might be something we should add anyway, even if we add custom file mark, since this feature might be useful to decent chunk of ytt users.

pivotaljohn replied:
> i do like this approach as well. this might be something we should add anyway, even if we add custom file mark, since this feature might be useful to decent chunk of ytt users.

That's fair — especially given that once Schema are in place, the Configuration Author will _almost always_ define data values via _that_ mechanism leaving Data Values as mostly Configuration Consumer customizing their `ytt` run.. so _that_ file will understandably _not_ likely in the same directory structure as the ytt library they are using. 👍🏼

pivotaljohn replied:
🤔 

However, it _does_ introduce non-trivial interface complexity:
- there would be two ways to indicate a plain YAML file containing Data Values documents
- there would be two ways to include files for processing (what does it mean to include a file both via a `-f` path _and_ `--data-values-file`?)

The convenience of this feature needs to outweigh the cost of that complexity.

Would it?

pivotaljohn replied:
I captured this as a "Part Two" to the specification, for now.

cppforlife commented on proposals/ytt/002-raw-data-values/README.md:
```diff
@@ -0,0 +1,254 @@
+
+`ytt` has an existing mechanism for indicating what type of file a given input is: [file marks](https://carvel.dev/ytt/docs/latest/file-marks/).
+
+Meet the specified needs by introducing another file type: `yaml-data-values`.
```
i like this approach.

one deficiency i just thought of with file marks is that they operate on relative path names.

```
config/
  config.yml
  values.yml
custom/
  values.yml
```

when you run `ytt -f config/ -f custom/ --files-inspect` you get

```
- config.yml
- values.yml
- values.yml
```

in this case, setting file mark would affect both values.ymls. this seems like a file mark deficiency but i think it would have to be solved before we can proceed with this.

pivotaljohn replied:
I see this as _exclusively_ a "feature" of the File Mark mechanism. Whatever is done to "fix" resolving those file references would address this concern.

The risk of deferring is low given how rare this case is:
- the files must be identically named,
- appear at the same relative path, and
- one of the files should not be treated as a Data Values file.

I submit that, in practice, this concern — while it ought to be addressed — is severable from this proposal.

pivotaljohn replied:
Captured this as "Open Question" **Q2** in the proposal.

cppforlife replied:
> the files must be identically named,

values.yml is pretty common name.

> appear at the same relative path, and

that would be common in kapp-controller case.

> one of the files should not be treated as a Data Values file.

this is where it gets a little weird, values.yml within templates will have #@data/values annotation (written by package authors). values.yml provided by the user will not have that annotation. will use of file mark raise an error on files that do have that annotation?

> Whatever is done to "fix" resolving those file references would address this concern.

this feels like a prereq to safe usage of this feature. a possibility of collision doesnt really inspire confidence in this feature imho.

pivotaljohn replied:
> that would be common in kapp-controller case.

huh... what does that look like? ... er, in what way does `kapp-controller` usage make this really common?

pivotaljohn replied:
> will use of file mark raise an error on files that do have that annotation?

[It's a no-op](https://github.com/vmware-tanzu/carvel-community/pull/21/files#diff-493463cfb65c2b66469184b92246b731ee03b4d0dd3ed7837cd914f725e97141R197
). Which specifically means that the file will be processed without alteration.

pivotaljohn replied:
Created an "Idea" "Discussion": vmware-tanzu/carvel-ytt#326 to work out a solution for disambiguating files while applying marks.

cppforlife replied:
> huh... what does that look like? ... er, in what way does kapp-controller usage make this really common?

im thinking that most folks per our recommendation would have values.yml included in their package, but also user installing package would probably pick a familiar name like values.yml as a key in Secret.

> It's a no-op. Which specifically means that the file will be processed without alteration.

hmm. is that a good thing it's a noop? seems like user is trying to do same thing twice, imho a bit weird. any reason not to error?

pivotaljohn replied:
> hmm. is that a good thing it's a noop? seems like user is trying to do same thing twice, imho a bit weird. any reason not to error?

I see the annotation and the file mark as semantic synonyms... so they are merely repeating themselves: they _really really_ want that file to be treated as a Data Values.

When `ytt` is being used by an upstream tool (e.g. `kapp-controller` servicing an `App`), it's quite useful to be more permission in accepting input here. It would unnecessarily limiting to _require_ the end user to _not_ supply an annotated Data Values, merely because it was being processed by a tool that file marked the same input as such.

Can you articulate the misunderstanding the user might have when specifying both the annotation _and_ the file mark that would be cause a surprising result for them?

pivotaljohn replied:
This line of thinking became moot when we changed direction. The input to `--data-values-file` specifically rejects inputs that are annotated.

Resolving.

cppforlife commented on proposals/ytt/002-raw-data-values/README.md:
```diff
@@ -0,0 +1,254 @@
+- https://github.com/vmware-tanzu/carvel-ytt/issues/81
+- https://github.com/vmware-tanzu/carvel-ytt/issues/51
+
+### Scenario: Consumer Configuration is Additive
```
there is an interesting case of providing complex values. e.g. if data value accepts a map, how would we deal with new map keys being merged? (currently overlay library would complain that it hasnt seen that key).

e.g.

```
annotations:
  foo: blah
```

pivotaljohn replied:
>  how would we deal with new map keys being merged?

I'm thinking steady-state:
- for this feature, we'd relax the overlay (effectively `@overlay/match-child-default missing_ok=True`)
- that schema will be in play ... and _that_ mechanism is used to constrain input.

Do you see any gaps in there?

pivotaljohn replied:
I meant to include a use-case where the targeted file _contains_ annotations.

In that case, this file mark should be a no-op. So, if the Configuration Consumer marks the contents as `@data/values` without the `@overlay/match-child-default missing_ok=True` and there is no schema in play, they'll get that overlay key-name check back.

Will add that.

jtigger committed:
Revised based on March 10 review

- promoted `--data-values-file` approach as a top-level approach.
  - articulated the attendant optional complexity involved.
- marked files will be annotated "missing ok" as well.
- marked files that contain annotations are a no-op.
- added two open questions

pivotaljohn commented:
Question came up in meatspace. Not exclusively, but especially with respect to motivating the `--data-values-file` flag option:

Q: How would you provide Data Values (without `ytt` annotations) that you do _not_ wish to be written to disk (e.g. because they contain sensitive data)?

jtigger committed:
Incorporate feedback and progress made on supporting work

- Answered Q2
- refined analysis of the complexity involved in including the
`--data-values-file` flag.
- ... a bunch of wordsmithing

jtigger committed:
Rewrite proposal to use the --data-values-file flag

... on further discussion, we realized that this approach would make far
more simple (and compact) the implementation of accepting Data Values).

- move proposal back into "draft" status; digesting all the
ramifications of this move is still a work-in-progress.

jtigger committed:
Flesh out many sections from rewrite

- start to include a body of examples on which the prose will leverage.

cppforlife commented on proposals/ytt/002-raw-data-values/README.md:
```diff
@@ -0,0 +1,351 @@
+where:
+- `libref` directs the values to the specified private library
+  - format: `@(<library-name> | ~<alias>):` (i.e. identical to the syntax for other `--data-value...` flags)
+- `+:` if present indicates that the Data Values in the document(s) should be set whether or not they have been previously declared.
```
i think we should make all data value keys be effectively missing_ok=True and rely on schema to deal with checking structure.

pivotaljohn replied:
How should it behave when there is not schema present?

cppforlife replied:
i would say the same. since we would encourage everybody to use schemas for safety im less concerned about it. it also does open up door for some folks who want to use -v foo=blah without specifying schemas or data values overlays.

jtigger committed:
Complete full set of examples

- add tables of content
- add references to examples from within the prose.
- include requirements for "clear and consistent messaging."

cppforlife commented on proposals/ytt/002-raw-data-values/README.md:
```diff
@@ -0,0 +1,351 @@
+
+In short:
+- introduce a new flag: `--data-values-file`;
+- accept only plain YAML;
```
what do we do with comments in those files? may be disallow #@?

cppforlife commented on proposals/ytt/002-raw-data-values/README.md:
```diff
@@ -0,0 +1,351 @@
+
+#### Multiple Documents
+
+`--data-values-file` can be specified multiple times. The file supplied in as a parameter can contain zero or more documents.
```
left most being applied first.

pivotaljohn replied:
Excellent. Added clarification.

cppforlife commented on proposals/ytt/002-raw-data-values/README.md:
```diff
@@ -0,0 +1,351 @@
+
+#### Miscellaneous
+
+- YAML comments (i.e. strings prefixed with `#`) are permitted in Data Values files given they are treated as plain YAML and not a `ytt` template.
```
may be disallow #@?

pivotaljohn replied:
Meant to include this. Now added (with rationale).

cppforlife commented on proposals/ytt/002-raw-data-values/README.md:
```diff
@@ -0,0 +1,351 @@
+#### Interaction with Schema
+
+There are no new interactions between Data Values supplied with this new flag and schema. 
+- Data Values supplied with this new flag are typed, checked, and validated against schema in exactly the same way as Data Values specified via the other `--data-value` flags.
```
are we validating and type checking as we apply each doc from --data-values-file?

pivotaljohn replied:
Yes. in terms of implementation details, each document in a `--data-values-file` is converted into an overlay and added to the list of overlays to apply during Data Values Pre Processing — no different than from any other source.

cppforlife commented on proposals/ytt/002-raw-data-values/README.md:
```diff
@@ -0,0 +1,351 @@
+
+#### Example: Scalar Data Value
+
+_While rare, it is valid to specify a single data value that is a scalar._
```
hehe. 👍

pivotaljohn replied:
😏

jtigger committed:
Address: preventing ytt annotations in Data Values Files

- more precisely state the order in which documents from Data Values
Files are applied.

jtigger committed:
Transition (back) to: in-review

Rewrite is now draft complete.

cppforlife commented on proposals/ytt/002-raw-data-values/README.md:
```diff
@@ -0,0 +1,478 @@
+As such:
+- YAML comments (i.e. strings prefixed with `#`) are permitted (as well as other features enjoyed by plain YAML)
+- `ytt` annotations (i.e. strings prefixed with `#@`) are _not_ permitted
+  - this prevents end-users from including executable bits, making it possible to better secure integration with `ytt`
```
since it would be plain yaml, we wouldnt even interpret #@. i think rationale for disallowing it is that it would catch cases where somebody mistakenly placed #@ into data-values-file and thought it would execute but it would be dropped.

pivotaljohn replied:
Makes sense... will put this as the primary rationale.

cppforlife commented on proposals/ytt/002-raw-data-values/README.md:
````diff
@@ -0,0 +1,478 @@
+Given how common this flag will be used, it also has a shorthand form:
+
+```console
+ytt ... -d [<libref>][+:]<filepath>
````
any alternative short flag names?

pivotaljohn replied:
no. you got any?

ewrenn8 commented on proposals/ytt/002-raw-data-values/README.md:
```diff
@@ -0,0 +1,478 @@
+
+#### Clear Consistent Messaging Around Data Values
+
+For the features described in this proposal to be well adopted and not create confusion, clear and consistent messaging needs to be crafted.
```
I am not a huge fan of this naming and think, when taken in combination with error messages, it could add confusion. If my Data Values Files fail to combine nicely, I will get an overlay error message. When looking at the docs I will see both Data Values Files and Data Values Overlays. I would likely go for overlays in this case since the error message is about overlaying, but that would be the wrong spot to look.

This assumes that error messages around data values haven't been updated.

ewrenn8 commented on proposals/ytt/002-raw-data-values/README.md:
```diff
@@ -0,0 +1,478 @@
+
+As such:
+- YAML comments (i.e. strings prefixed with `#`) are permitted (as well as other features enjoyed by plain YAML)
+- `ytt` annotations (i.e. strings prefixed with `#@`) are _not_ permitted
```
Is there some rationale for this thats been discussed outside this proposal?

I'm not sure I understand why this flag can't just be like -f but everything given to it is automatically added to the heap of data values files. This may be another iteration of the flag, but introducing the added overhead of knowing if I need to provide my data values file via -f or -d based on its contents might get kind of annoying and surprising to existing users that read the flag as a way to specify anything data values related. It feels like we are creating another conceptual data values file (Plain Data Values YAML), which further complicates already complicated concepts.

jtigger committed:
Name concerns around confusion from splitting data values

- Added the concerns:
  - Implies a Significant Effort to Implement
  - Potentially Confusing that Data Values Overlays are not Data Values files
- Include, describe, and analyze the previous approach (using file
    marks)
- Remove the `+:` operator, favoring always allowing data values through
this mechanism to add new values.
- Added Open Question about whether the recognized complexity is worth
the upsides of this approach.
- Extracted commentary into a separate section: "Analysis" to keep clear
where the spec ends and the evaluation begins.

(minor edits)
- fix typos/errors in some examples

jtigger committed:
Add minor yet clarifying notes

originated from an in-person review with Eli and Dmitriy.

jtigger committed:
Incorporate rationale hatched in proposal discussion

- approvers (Eli & Dmitriy) gave feedback and worked through the open
question in a live session. Proposal now (hopefully) incorporates those
perspectives.

jtigger committed:
moved to "accepted"
